http/src/lib.rs

```diff
@@ -254,76 +236,22 @@ impl<M: jsonrpc::Metadata, S: jsonrpc::Middleware<M>> ServerBuilder<M, S> {
 			extractor: self.meta_extractor,
 		};
 
-		let (l, srv) = hyper::Server::http(addr)?.handle(move |control| {
-			let hosts = hosts.lock().clone();
-			ServerHandler::new(
-				jsonrpc_handler.clone(),
-				cors_domains.clone(),
-				hosts,
-				request_middleware.clone(),
-				control,
-			)
-		})?;
-
-		// Add current host to allowed headers.
-		// NOTE: we need to use `l.addrs()` instead of `addr`
-		// it might be different!
-		{
-			let mut hosts = hosts_setter.lock();
-			if let Some(current_hosts) = hosts.take() {
-				let mut new_hosts = current_hosts.into_iter().collect::<HashSet<_>>();
-				for addr in l.addrs() {
-					let address = addr.to_string();
-					new_hosts.insert(address.clone().into());
-					new_hosts.insert(address.replace("127.0.0.1", "localhost").into());
-				}
-				// Override hosts
-				*hosts = Some(new_hosts.into_iter().collect());
-			}
-		}
+		let server_handler = ServerHandler {
+			jsonrpc_handler: jsonrpc_handler,
+			cors_domains: self.cors_domains,
+			allowed_hosts: self.allowed_hosts,
+			middleware: self.request_middleware,
+		};
 
-		let handle = thread::spawn(move || {
-			srv.run();
-		});
+		let listening = hyper::Server::http(addr)?.handle(server_handler)?;
 
 		Ok(Server {
-			server: Some(l),
-			handle: Some(handle),
-			remote: Some(eloop),
+			server: Some(listening),
 		})
 	}
 }
 
 /// jsonrpc http server instance
 pub struct Server {
```

Comment:
I'm afraid those methods do not exist any more. There were some many api changes in hyper, that instead of removing the code line by line, I just started from scratch and begin to pull your code there.

Nevertheless, if I continue with this pr, I will try to address these problems ;)